docs/source/blueprints/BP003.rst

```diff
@@ -0,0 +1,366 @@
+now limited to tasks being prepared and ready to be started.
+
+Supposing that the "status uri" 127.0.0.1:8080, was set by the job, as its
+internal status server, it must be started before any task, to avoid any status
+message being lost.
+
+At this stage, Task #2 is started, and Task #1 is now ready.
+
+Iteration IV
+~~~~~~~~~~~~
+
+On the fourth iteration, Task #1 is started.
+
+Note: the ideal level of parallelization is still to be defined, that is, it may
+be that triaging and preparing and starting tasks, all run concurrently. An
+initial implementation that, on each iteration, looks at all Task states, and
+attempts to advance them further, blocking other Tasks as much as little as
+possible should be acceptable.
+
+Iteration V
+~~~~~~~~~~~
+
+On the fifth iteration, the spawner reports that Task #2 is not alive anymore,
+and the status server has received a message about it (and also a message about
+Task #1 having started).
+
+Because of that, Task #2 is now considered ``FINISHED``.
+
+And Task #1 is still a ``STARTED`` task.
+
+Final Iteration
+~~~~~~~~~~~~~~~
+
+After several iterations with no status changes, and because of a timeout
+implementation at the job level, it's decided that Task #1 is not to be waited
+on.
+
+The spawner continues to inform that Task #1 is alive (from its PoV), but no
+further status message has been received. Provided the spawner has support for
+that, it may attempt to clean up the task (such as destroying a container or
+killing a process). In the end, it's left with.
+
+Tallying results
+~~~~~~~~~~~~~~~~
+
+The nrunner plugin should be able to provide meaningful results to the Job, and
+consequently, to the user, based on the resulting information on the final
+iteration.
+
+Notice that some information will come, such as the ``PASS`` for the first test,
+will come from the "result" given in a status message from the task itself. Some
+other status, such as the ``INTERRUPTED`` status for the second test will not
+come from a status message received, but from a realization of the actual
+management of the task execution. It's expected to other information will also
+have to be inferred, and "filled in" by the nrunner plugin implementation.
+
+In the end, it's expected that results similar to this would be presented::
+
+    JOB ID     : f59bd40b8ac905864c4558dc02b6177d4f422ca3
+    JOB LOG    : /home/cleber/avocado/job-results/job-2020-05-20T17.58-f59bd40/job.log
+     (1/2) tests.py:Test.test_2: PASS (2.56 s)
+     (2/2) tests.py:Test.test_1: INTERRUPT (900 s)
+    RESULTS    : PASS 1 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 1 | CANCEL 0
+    JOB TIME   : 0.19 s
+    JOB HTML   : /home/cleber/avocado/job-results/job-2020-05-20T17.58-f59bd40/results.html
+
+Notice how Task #2 shows up before Task #1 because it was both started first and
+finished earlier. There may be issues associated with the current UI to deal
+with regarding out of order task status updates.
```

Comment:
> So, unless I'm missing something, you have explained a proposal "task life cycle", with some iteration examples. That is great and for sure is needed here. I think that this life cycle is a good starting point for us to work with this scheduler idea.
> 
> But this is the _what_ part of the scheduler. Yes, you described _how_ each task will move between states, but I'm still missing here the an important _how_ part: How do you pretend to implement this? I'm not talking about the draft code that you have, but just some basic idea of how do you pretend to code this?
> 

I pretend to code this as straightforward as with as few unnecessary abstractions as possible at this stage, hence the draft code.  So, what I have in mind at this point, is one entity that knows about the state of all tasks, on the different stages of their life cycle, and pushes them in the appropriate direction.  Basically a state machine manager.

It should be clear that I honestly don't think the _how_ is *as important as* the _what_ for this specific endeavor at this stage.  I mentioned elsewhere that I started using `asyncio` as a way to avoid blocking on operations, but any other sound implementation achieving the same goals would be valid.  This general feeling is the same I have about the other implementation details.

> I can think of a few architectures/ways to do this, and depending on the way, maybe some "states" could be omitted.
> 
> So, basically I do have a few questions so far:
> 
> 1. What have you thought so far about the components of this (in words/draws, please ;) )?

I think the following components *could* end up existing:

* task triager
* task requirements fulfifiller
* task executor
* task monitor
* rogue task killer

But like I said before, I don't think these components *have* to exist as standalone separate entities.  We're not planning a task scheduler to be consumed and augmented as a technology at this point.

> 2. Are we going to have a watchdog to monitor the task' state?

Sure, monitoring the tasks is needed to:

 * Give the user update on the task progress (like the current human UI does)
 * Implement test timeouts
 * Implement job timeouts
 * Collect results

> 3. How the components are going to talk to each other? Pub/Sub? Message queue?

This is the crux of the issue.  I don't believe separate components are a requirement at this point.  Not only that, making them separate, standalone components, leads to this type of question.  Again, I'm proposing a Linux instead of a Minix at this stage. 

> 4. Runners are going to be ready for running tasks a priori? Or is it the same as today?

I don't think I understood your question.  Are you asking if the current runners we have should be suitable to run tasks in the current proposal?  If so, then, yes, they should be capable to be used as they are (or with minor adjustments). 

> 5. I'm assuming that we are going to have something like `task.state` with an enum pointing to one of those states that you described. Is that correct? Or are you thinking in something different?

I think I mentioned that the task state could either be represented as an internal attribute, or simply by moving them to the appropriate queue/stage.  I favor the second approach, maybe because I remember how the Linux process scheduler achieved simplicity and O(1) by having queues ready to be used, instead of processing task data every time a scheduling decision needed to be made.

> 6. This seems to be a perfect scenario for a producer-consumer architecture (maybe a Pub/Sub or Message Bus could be helpful here and easy to implement). Is that what you have in mind?
> 

I think I would make more mistakes at this point if I tried to change my implementation bias into a Pub/Sub or Message Bus approach.  I will think a little bit more about that, but I can't promise that I trust myself into "forcing" a type of implementation and succeeding.

> I know that we have today a very simple mechanism that sends the tasks over a command. But maybe this is a perfect time to think about this. Because, IMO after we implement this, changing things in the future could be difficult.

This is one area that I don't think will change much, if at all.  I agree that the overall design changes wrt runners, tasks, etc should be minimized at this point, because we expect users to be able to write their own runners for their own test types (and code those types into hint files, for instance).  But how Avocado internally will execute them, and monitor their execution, which is the scope of this work, should be completely invisible to those users.  So changes in this "scheduler" could happen post-LTS without disruption to users IMO.

Reply:
> > So, unless I'm missing something, you have explained a proposal "task life cycle", with some iteration examples. That is great and for sure is needed here. I think that this life cycle is a good starting point for us to work with this scheduler idea.
> > But this is the _what_ part of the scheduler. Yes, you described _how_ each task will move between states, but I'm still missing here the an important _how_ part: How do you pretend to implement this? I'm not talking about the draft code that you have, but just some basic idea of how do you pretend to code this?
> 
> I pretend to code this as straightforward as with as few unnecessary abstractions as possible at this stage, hence the draft code. So, what I have in mind at this point, is one entity that knows about the state of all tasks, on the different stages of their life cycle, and pushes them in the appropriate direction. Basically a state machine manager.
> 
> It should be clear that I honestly don't think the _how_ is _as important as_ the _what_ for this specific endeavor at this stage. I mentioned elsewhere that I started using `asyncio` as a way to avoid blocking on operations, but any other sound implementation achieving the same goals would be valid. This general feeling is the same I have about the other implementation details.

Yes, I understand that "why" and "what" are the most important part here, but IMO we still could have a very basic "how". 

The whole idea of a blueprint is: "Here is the problem, here is what we can do, and this is how I thought about solving this problem. Makes sense? Or should I go in a different direction?" 
 
> > I can think of a few architectures/ways to do this, and depending on the way, maybe some "states" could be omitted.
> > So, basically I do have a few questions so far:
> > 
> > 1. What have you thought so far about the components of this (in words/draws, please ;) )?
> 
> I think the following components _could_ end up existing:
> 
>     * task triager
> 
>     * task requirements fulfifiller
> 
>     * task executor
> 
>     * task monitor
> 
>     * rogue task killer
> 
> 
> But like I said before, I don't think these components _have_ to exist as standalone separate entities. We're not planning a task scheduler to be consumed and augmented as a technology at this point.

Even the executor? I do have some difficulties to accept the whole thing as one coupled thing. I can see most of them as one piece but I'm always trying to put the executor in a different machine.
 
> > 1. Are we going to have a watchdog to monitor the task' state?
> 
> Sure, monitoring the tasks is needed to:
> 
>     * Give the user update on the task progress (like the current human UI does)
> 
>     * Implement test timeouts
> 
>     * Implement job timeouts
> 
>     * Collect results

Good, would be nice to have this inside this BP.
 
> > 1. How the components are going to talk to each other? Pub/Sub? Message queue?
> 
> This is the crux of the issue. I don't believe separate components are a requirement at this point. Not only that, making them separate, standalone components, leads to this type of question. Again, I'm proposing a Linux instead of a Minix at this stage.

Please, don't see that way. I'm not trying to push this. I'm just trying to understand the problem and *how* you are trying to solve, to see how this compares with a pub/sub solution or other solution. Again, I'm missing the hows.
  
> > 1. Runners are going to be ready for running tasks a priori? Or is it the same as today?
> 
> I don't think I understood your question. Are you asking if the current runners we have should be suitable to run tasks in the current proposal? If so, then, yes, they should be capable to be used as they are (or with minor adjustments).

Sorry, my question was: Are we going to wakeup (start) the runners or they are going to be there, listening and ready for tasks? I remember that you mention this at some point in the past but I'm not sure if this is one target here.

> > 1. I'm assuming that we are going to have something like `task.state` with an enum pointing to one of those states that you described. Is that correct? Or are you thinking in something different?
> 
> I think I mentioned that the task state could either be represented as an internal attribute, or simply by moving them to the appropriate queue/stage. I favor the second approach, maybe because I remember how the Linux process scheduler achieved simplicity and O(1) by having queues ready to be used, instead of processing task data every time a scheduling decision needed to be made.

I'm in favor of different queues too, but we could have both: change the state and move to the queue. Or the state is a `@property` that checks what is the current queue. 
 
> > 1. This seems to be a perfect scenario for a producer-consumer architecture (maybe a Pub/Sub or Message Bus could be helpful here and easy to implement). Is that what you have in mind?
> 
> I think I would make more mistakes at this point if I tried to change my implementation bias into a Pub/Sub or Message Bus approach. I will think a little bit more about that, but I can't promise that I trust myself into "forcing" a type of implementation and succeeding.

I think that I'm repeating myself already, and sorry if I sound like a "broken record player", but IMO the purpose of a blueprint or RFC is not only to discuss the problem itself, but also as many as possible solutions we think, and avoid regrets in the future.
 
> > I know that we have today a very simple mechanism that sends the tasks over a command. But maybe this is a perfect time to think about this. Because, IMO after we implement this, changing things in the future could be difficult.
> 
> This is one area that I don't think will change much, if at all. I agree that the overall design changes wrt runners, tasks, etc should be minimized at this point, because we expect users to be able to write their own runners for their own test types (and code those types into hint files, for instance). But how Avocado internally will execute them, and monitor their execution, which is the scope of this work, should be completely invisible to those users. So changes in this "scheduler" could happen post-LTS without disruption to users IMO.

I'm just trying to understand better the problem and your solution, to make sure that we are making the best choice. Maybe this is clear in your mind and would be nice to see this written here in this BP. As I said, we don't need to implement everything here. But document as much as possible, at this stage, could be helpful.

I would like much to see this discussion inside the BP so we can have the thoughts here organized and for future reading.

Thank you so much for this discussion.